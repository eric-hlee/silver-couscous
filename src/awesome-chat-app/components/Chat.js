import React from 'react'
import SentMessage from './SentMessage'
import RepliedMessage from './RepliedMessage'
import { TEXTAREA_PLACEHOLDER } from '../constants/ActionTypes'


class Chat extends React.Component {

  constructor(props) {
    super(props);
    this.messagesEnd = React.createRef();
  }

  scrollToBottom = () => {
    this.messagesEnd.scrollIntoView({ behavior: "smooth" });
  }

  componentDidMount() {
    this.scrollToBottom();
  }

  componentDidUpdate() {
    var el = this.refs.message;
    el.scrollTop = el.scrollHeight;
    this.scrollToBottom();
  }

  render() {

    const style_conversation = {
      display: 'flex',
      height: '100%',
      flexDirection: 'column',
    }

    const style_message = {
      height: '500px',
      marginBottom: '20px',
      overflow: 'auto',
    }

    const style_chatBox = {
      display: 'inline-block',
      width: '50%',
      position: 'relative',
      padding: '10px',
    }

    const style_textarea = {
      width: '100%',
      padding: '1em',
      boxSizing: 'border-box',
      border: '2px solid #ccc',
      borderRadius: '4px',
      fontSize: '16px',
      minHeight: '58px',
      resize: 'none',
      outline: 'none',
      overflow: 'auto',
    }

    return (
      <div className="Conversation" style={style_conversation}>
        <div className="Message" ref="message" style={style_message}>
          {this.props.messageList.map(message => {
            if (message.fromUser) {
              return (
                <SentMessage
                  key={message.id}
                  date={message.time}
                  text={message.message}
                />
              );
            } else {
              return (
                <RepliedMessage
                  key={message.id}
                  date={message.time}
                  text={message.message}
                />
              );
            }
            }
          )}
          <div ref={(el) => { this.messagesEnd = el; }} />
        </div>
        <div className="ChatBox" style={style_chatBox}>
          <div className="Textarea">
            <textarea id="replyTextArea" type="text" autoFocus
              value={this.props.lastInput}
              placeholder={TEXTAREA_PLACEHOLDER} 
              onKeyDown={this.props.onEnter} onChange={this.props.handleChange}
              style={style_textarea}
            />
          </div>
        </div>
      </div>
    );
  }
}

export default Chat