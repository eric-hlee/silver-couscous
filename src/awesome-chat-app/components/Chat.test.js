import React from 'react'
import { mount } from 'enzyme'
import Chat from './Chat'

describe('Chat', () => {
  it('should render self and subcomponents', () => {
    const props = {
      messageList: [
        {
          id: 0,
          time: '2019-06-07T11:22:05.032Z',
          message: 'Welcome to the chat',
          fromUser: false
        }
      ],
      messagesEnd: <div />
    }
    Element.prototype.scrollIntoView = jest.fn();
    const wrapper = mount(<Chat {...props} />)
    expect(wrapper).toMatchSnapshot()
  })
})
