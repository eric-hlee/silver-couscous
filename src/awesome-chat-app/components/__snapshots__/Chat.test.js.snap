// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should render self and subcomponents 1`] = `
<Chat
  messageList={
    Array [
      Object {
        "fromUser": false,
        "id": 0,
        "message": "Welcome to the chat",
        "time": "2019-06-07T11:22:05.032Z",
      },
    ]
  }
  messagesEnd={<div />}
>
  <div
    className="Conversation"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
      }
    }
  >
    <div
      className="Message"
      style={
        Object {
          "height": "500px",
          "marginBottom": "20px",
          "overflow": "auto",
        }
      }
    >
      <RepliedMessage
        date="2019-06-07T11:22:05.032Z"
        key="0"
        text="Welcome to the chat"
      >
        <div
          className="MessageBubbleGroup"
          style={
            Object {
              "overflow": "auto",
              "position": "relative",
            }
          }
        >
          <div
            className="RepliedMessageBubble"
            style={
              Object {
                "border": "1px solid #555555",
                "borderRadius": "10px",
                "display": "inline-block",
                "marginBottom": "15px",
                "padding": "15px",
              }
            }
          >
            <div
              className="Date"
            >
              <span>
                Sent on 
                2019-06-07T11:22:05.032Z
              </span>
            </div>
            <div
              className="Message"
              style={
                Object {
                  "fontFamily": "Helvetica Neue, Arial, sans-serif",
                  "fontSize": "16px",
                  "marginTop": "6px",
                }
              }
            >
              <span>
                Welcome to the chat
              </span>
            </div>
          </div>
        </div>
      </RepliedMessage>
      <div />
    </div>
    <div
      className="ChatBox"
      style={
        Object {
          "display": "inline-block",
          "padding": "10px",
          "position": "relative",
          "width": "50%",
        }
      }
    >
      <div
        className="Textarea"
      >
        <textarea
          autoFocus={true}
          id="replyTextArea"
          placeholder="Type your message here and press enter to send..."
          style={
            Object {
              "border": "2px solid #ccc",
              "borderRadius": "4px",
              "boxSizing": "border-box",
              "fontSize": "16px",
              "minHeight": "58px",
              "outline": "none",
              "overflow": "auto",
              "padding": "1em",
              "resize": "none",
              "width": "100%",
            }
          }
          type="text"
        />
      </div>
    </div>
  </div>
</Chat>
`;
